src/crimson/osd/heartbeat.cc

```diff
@@ -309,6 +312,34 @@ seastar::future<> Heartbeat::handle_you_died()
   return seastar::now();
 }
 
+void Heartbeat::heartbeat_check()
+{
+  failure_queue_t failure_queue;
+  const auto now = clock::now();
+  for (const auto& [osd, peer_info]: peers) {
+    if (clock::is_zero(peer_info.first_tx)) {
+      continue;
+    }
+
+    if (peer_info.is_unhealthy(now)) {
+      logger().error(" heartbeat_check: no reply from osd.{} "
+		     "since back {} front {} (oldest deadline {})",
+		     osd, peer_info.last_rx_back, peer_info.last_rx_front,
+		     peer_info.ping_history.begin()->second.deadline);
+      failure_queue[osd] = std::min(peer_info.last_rx_back,
+				    peer_info.last_rx_front);
+    }
+  }
+  if (!failure_queue.empty()) {
+    // send_failures can run in background, because messages
+    // are sent in order, if later checks find out the previous
+    // "failed" peers to be healthy, that "still alive" messages
+    // would be sent after the previous "osd failure" messages
+    // which is totally safe.
```

Comment:
actually i am more concerned with the life cycle management of the captured `Heartbeat` instance. as once it's captured by a continuation and sent to reactor we have no way to ensure it outlives the capturing continuation, in this case, when `timer` gets triggered, it schedules a continuation to be executed by reactor. but OSD **could** be destroyed before the last chained continuation of the scheduled event is executed by reactor even the timer is removed from reactor by `~timer()`. that's why the tasks scheduled by timer is either atomic or just a single continuation or multiple continuations protected by gate or a semaphore.

Reply:
Ah, thanks:-)

Reply:
@xxhdx1985126 could you please address my comment?

Reply:
added comment, please take a look at [34017](https://github.com/ceph/ceph/pull/34017):-)

Reply:
thank you! will review it shortly.

Reply:
You're welcome:-)